Add tracking for the contact form block

scruffian opened the pull request:
Right now we can track how many contact forms are submitted, but it's not possible to distinguish which ones are from the shortcode and which ones are from the block. This adds a tracks event specifically for the block, so that we can track it independently of the shortcode.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Relies on the `hasFormSettingsSet` attribute which is set by the block
* When this attribute exists we add a hidden field to the form with the name `is_block`
* If this hidden field is submitted we then add a value to the `extra_values` array
* If `is_block` is set in the `extra_values` array, then send a `jetpack_contact_form_gutenberg_block_message_sent` event to Tracks

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Adding features to Jetpack

#### Testing instructions:
* Create a new post and add a contact form block
* From the front end of your site, submit the contact form
* Go to Tracks and search for the event `jetpack_contact_form_gutenberg_block_message_sent`
* Now create another post and add the contact form shortcode
* From the front end of your site, submit the contact form
* Go to Tracks and search for the event `jetpack_contact_form_gutenberg_block_message_sent` - you should not see it.

scruffian committed:
Add tracking for the contact form

scruffian committed:
Add file

scruffian committed:
Make it possible to track contact form submissions only coming from Gutenberg blocks

jeherve commented on modules/contact-form/tracks-events.php:
```diff
@@ -0,0 +1,15 @@
+
+use Automattic\Jetpack\Tracking;
+
+function tracks_record_grunion_pre_message_sent( $post_id, $all_values, $extra_values ) {
```
Should this function use a `jetpack_` prefix?

scruffian replied:
👍

jeherve commented on modules/contact-form/tracks-events.php:
```diff
@@ -0,0 +1,15 @@
+function tracks_record_grunion_pre_message_sent( $post_id, $all_values, $extra_values ) {
+	$tracking = new Automattic\Jetpack\Tracking();
+	if ( $extra_values['is_block'] ) {
+		$tracking->tracks_record_event( wp_get_current_user(), 'jetpack_contact_form_gutenberg_block_message_sent', array(
```
Do we need `gutenberg` in the event name here? Wouldn't `block` mean that this comes from a form block?

scruffian replied:
No, good call. I have removed it.

jeherve commented on modules/contact-form/tracks-events.php:
```diff
@@ -0,0 +1,15 @@
+	$tracking = new Automattic\Jetpack\Tracking();
+	if ( $extra_values['is_block'] ) {
+		$tracking->tracks_record_event( wp_get_current_user(), 'jetpack_contact_form_gutenberg_block_message_sent', array(
+			'entry_permalink' => $all_values['entry_permalink'],
```
It may be worth escaping the data here and below before it gets sent, with `esc_url` and `absint` for example.

scruffian replied:
Good idea. Done!

jeherve reviewed (dismissed):
All tests are currently failing, so they'll need to be updated:
```
WP_Test_Grunion_Contact_Form::test_process_submission_will_store_a_feedback_correctly_with_default_form
Undefined index: is_block
```

*****

Are those events triggered even when the site is in development mode, when the site owner has not given their consent for tracking yet?

scruffian committed:
update event name

scruffian committed:
escape data sent to tracks

scruffian committed:
add a jetpack prefix to the tracks event function

scruffian commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -26,6 +26,8 @@ function grunion_contact_form_require_endpoint() {
 	require_once GRUNION_PLUGIN_DIR . 'class-grunion-contact-form-endpoint.php';
 }
 
+require_once GRUNION_PLUGIN_DIR . 'tracks-events.php';
```
We'll want the tracking on WPCOM too, so I think we can leave it, assuming it will still work. I guess Tracks probably works differently on WPCOM.

kraftbj replied:
We'll need to resolve this to have the Fusion test pass. TC is failing on this require. So, either a `WPCOM` conditional to load a different file or have `tracks-events.php` be friendly to both environments (plus adding to Fusion and committing the file separately before the changes to `grunion-contact-form.php`

scruffian replied:
I added a `WPCOM` conditional. I'm not sure how to update the wpcom patch or check the fusion tests...

scruffian committed:
add new file to the phpcs whitelist

scruffian commented on modules/contact-form/tracks-events.php:
```diff
@@ -0,0 +1,15 @@
+<?php
```
Done

scruffian commented on modules/contact-form/tracks-events.php:
```diff
@@ -0,0 +1,15 @@
+
+function tracks_record_grunion_pre_message_sent( $post_id, $all_values, $extra_values ) {
+	$tracking = new Automattic\Jetpack\Tracking();
+	if ( $extra_values['is_block'] ) {
```
We already track the submissions of contact forms, as these are synced with wpcom through the feedback custom post type. The problem we have is that we can't determine whether these submissions are coming from a block or a shortcode. More context here: pb5gDS-62-p2

scruffian committed:
fix undefined index error

scruffian commented:
> Are those events triggered even when the site is in development mode, when the site owner has not given their consent for tracking yet?

The contact form won't work unless Jetpack is connected, so we are safe in that respect.

jeherve commented:
> The contact form won't work unless Jetpack is connected

Jetpack's Contact Form is one of the features that is accessible when Jetpack is in [development mode](http://jetpack.com/support/development-mode), so we do need to take it into account imo.

scruffian committed:
fix undefined index error

scruffian committed:
dont send tracks events in development mode

scruffian commented on modules/contact-form/tracks-events.php:
```diff
@@ -0,0 +1,38 @@
+function jetpack_tracks_record_grunion_pre_message_sent( $post_id, $all_values, $extra_values ) {
+	$status = new Automattic\Jetpack\Status();
+
+	if ( ! $status->is_development_mode() && isset( $extra_values['is_block'] ) && $extra_values['is_block'] ) {
```
I added this check to make sure we don't send this Tracks events in development mode, but I think we should make that change more generally.

kraftbj replied:
With a quick look, I think you're right. We should add the dev mode check in this ballpark https://github.com/Automattic/jetpack/blob/a5a11059d24eb7ee50b25dbdb08f04ead55a0d50/packages/tracking/legacy/class.tracks-client.php#L54

scruffian replied:
I'll create a PR for this

scruffian replied:
Done here: https://github.com/Automattic/jetpack/pull/13698

scruffian committed:
use a different tracks implementation on wpcom

scruffian committed:
move wpcom tracking to a different file

jeherve committed:
Simplify approach to one single function

- keep that function in a file that is already synchronized with WordPress.com
- include both wpcom and Jetpack tracking approaches in that function.
- remove the dev mode check since this will be addressed in #13711

jeherve commented:
Ongoing discussions about this here:
p1570722249055600-slack-jetpack-crew

jeherve committed:
Send blog ID in recorded event when available

jeherve committed:
Feedback ID can include letters as well. Let's not sanitize as int

jeherve committed:
Specify a user when submission is anonymous

jeherve committed:
Use a different prefix for WordPress.com and Jetpack

If we don't do that we can't separate the 2 when looking at the data later on.

jeherve commented:
This should now be ready for a new review. The WordPress.com diff is also updated.

- We now include the blog ID when possible, to get valid Tracks events.
- We now fallback to the Jetpack Master User information when a contact form is submitted from a non-logged in user.
- We now have a different prefix for the event name that will be fired, depending on the platform. This will allow us to differentiate data coming from WordPress.com sites vs. Jetpack sites.

scruffian commented:
Thanks @jeherve these changes look great.

The only one I am not sure about is:

> We now have a different prefix for the event name that will be fired, depending on the platform. This will allow us to differentiate data coming from WordPress.com sites vs. Jetpack sites.

I worry that this might make it harder to get overall number for contact form submissions. @huguesvincent, what do you think?

huguesvincent commented:
Would it allow us to get the volume on DotCom alone? 
That's really what I was hoping to get here. I don't think we'll be considering to do any sort of tiring or carve a different experience depending on the volume you get. We might on DotCom particularly as ZBS integration comes along.

jeherve commented:
> Would it allow us to get the volume on DotCom alone?

Yes, you'd be able to track both Jetpack usage and WordPress.com usage separately, or together with a `%_contact_form_block_message_sent` wildcard.

scruffian commented:
>  or together with a %_contact_form_block_message_sent wildcard.

Oh I didn't know about this. That's great :+1:

jeherve committed:
Add missing comment *

jeherve reviewed (dismissed):
This works well in my tests, here and on wpcom, but I'd appreciate another review before merge since I contributed to this PR. 

Thank you!

dereksmart commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -3513,3 +3517,63 @@ function grunion_delete_old_spam() {
+		}
+
+		$tracking = new Automattic\Jetpack\Tracking();
+		$tracking->tracks_record_event( $event_user, $event_name, $event_props );
```
You may be able to instead use `record_user_event()` which will handle some of the logic you've written such as the blog_id and the `jetpack_` prefix to the event name.

jeherve replied:
Done in 012227f93e73f20984eec27d9f1096e929116375. I also updated the wpcom diff to match this.

jeherve committed:
Switch to using record_user_event

See https://github.com/Automattic/jetpack/pull/13688/files#r339803539

jeherve reviewed (approved):
This seems to work well in my tests. Merging.
